xADManagedServiceAccount: New resource for Managed Service Accounts

awickham10 opened the pull request:
Created a new MSFT_xADManagedServiceAccount resource for managing Managed Service Accounts (MSAs). 

Currently supported parameters:
- ServiceAccountName - Name of the service account
- Path - OU of MSA
- Ensure
- Description
- DisplayName
- Credential
- DomainController

As a note - I have not tested putting this in an actual DSC configuration, just did the testing that is included in the unit tests file.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xactivedirectory/196)
<!-- Reviewable:end -->

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/xActiveDirectory?pullRequest=196) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/xActiveDirectory?pullRequest=196) before we can accept your contribution.<br/><br/>:x: awickham10 [sign now](https://cla.opensource.microsoft.com/PowerShell/xActiveDirectory?pullRequest=196)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/xActiveDirectory?pullRequest=196) it.</sub>

johlju commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAx0IoA1Y2LN-lk7X0:-LDAx0IoA1Y2LN-lk7X1:b7smo8n) (no related file):*
Could you please run this resource through a manual integration test?

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAyRRn5zv0hxVNeINk:-LDAyRRn5zv0hxVNeINl:bj15bor) (no related file):*
Please add this new resource to the README.md.

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAyVSK4o8IUKbASs0a:-LDAyVSK4o8IUKbASs0b:b-jmcs32) (no related file):*
Could you please add an descriptive entry, for each change/issue, to the Unreleased section of the change log in the file README.md? If an entry resolves an issue please reference the issue. You may suffix (it's optional) each entry with your name. Example of the format for the entry
```
* Descriptive entry ([issue #123](https://github.com/PowerShell/xPSDesiredStateConfiguration/issues/123). [Name/Alias (@github_account)](https://github.com/github_account)`
```

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.mof, line 1 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAwanv7eo_Idv58EpX:-LDAwanv7eo_Idv58EpY:b8gp3bh) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.mof#L1)):*
> ```
> [ClassVersion("1.0.1.0"), FriendlyName("xADManagedServiceAccount")]
> ```

This file name should be suffixed with '.schema.mof'.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.mof, line 4 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAyHmhDymdGEa2jy0B:-LDAyHmhDymdGEa2jy0C:btxvomx) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.mof#L4)):*
> ```
> class MSFT_xADManagedServiceAccount : OMI_BaseResource
> {
>     [Key, Description("Name of the Active Directory managed service account")] String Name;
> ```

Should this be ServiceAccountName?

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 6 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAy9B64KuAxIs7_-H3:-LDAy9B64KuAxIs7_-H4:b-ogztn5) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L6)):*
> ```PowerShell
>     # culture="en-US"
>     ConvertFrom-StringData @'
>         AddingManagedServiceAccount           = Adding AD Managed Service Account '{0}'
> ```

Please end each localized message with full stop (`.`).

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 16 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAzBKY6ShqVZmBBuJT:-LDAzBKY6ShqVZmBBuJU:b-2x9pnd) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L16)):*
> ```PowerShell
> }
> 
> function Get-TargetResource
> ```

Please add comment-based help for each function. Throughout.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 30 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAzLMe8SGDLY0r69UV:-LDAzLMe8SGDLY0r69UW:bgl33y7) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L30)):*
> ```PowerShell
>         [ValidateNotNullOrEmpty()]
>         [System.String]
>         $Path,
> ```

We should not have any unused non-mandatory parameters in the Get-TargetResource function. Unused parameters should be removed from the Get-TargetResource.
https://github.com/PowerShell/DscResources/blob/master/BestPractices.md#get-targetresource-should-not-contain-unused-non-mandatory-parameters

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 66 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAzso0Bp43ljX6TDrn:-LDAzso0Bp43ljX6TDro:bdeanzw) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L66)):*
> ```PowerShell
>         $adServiceAccount = Get-ADServiceAccount @adServiceAccountParams -Property Name,DistinguishedName,Description,DisplayName
> 
>         $targetResource = @{
> ```

The hash table returned from the Get-TargetResource should include all properties in the schema.mof.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 76 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAznfj1pt6-tmuTL1z:-LDAznfj1pt6-tmuTL2-:b-dm75vq) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L76)):*
> ```
> ;
> ```

Semi-colon is not needed.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 79 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB26e_-hf9UVnoDCi6:-LDB26e_-hf9UVnoDCi7:buik1jc) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L79)):*
> ```PowerShell
>         }
>     }
>     catch [Microsoft.ActiveDirectory.Management.ADIdentityNotFoundException]
> ```

Maybe we should catch other problems that can occur too? Or, limit this try-catch to the call that should be cached?

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB-1fuAhBcYFep-9lG:-LDB-1fuAhBcYFep-9lH:b7a9w5x) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L82)):*
> ```PowerShell
>     {
>         Write-Verbose ($LocalizedData.ManagedServiceAccountNotFound -f $ServiceAccountName)
>         $targetResource = @{
> ```

Instead of duplicating the hash-table (more or less). Maybe we can add this to the beginning of the function, and if values are found they are set to the correct values?

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 109 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAzG0c1BcWCtmzzJeJ:-LDAzG0c1BcWCtmzzJeK:b-rnmcc7) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L109)):*
> ```PowerShell
> 
>         [Parameter()]
>         [ValidateSet("Present", "Absent")]
> ```

Please use single quote where applicable. Throughout.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 135 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB-zDcBuq1anFuiEu2:-LDB-zDcBuq1anFuiEu3:b-6nt93d) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L135)):*
> ```
> $targetResource
> ```

Could we call this `$getTargetResourceResult`? It's easier to read when reviewing with clear variable names to know what they contain. Throughout.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 251 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB1_Hb96WDpOMKRO02:-LDB1_Hb96WDpOMKRO03:buik1jc) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L251)):*
> ```PowerShell
>         }
>     }
>     catch [Microsoft.ActiveDirectory.Management.ADIdentityNotFoundException]
> ```

Maybe we should catch other problems that can occur too? Or, limit this try-catch to the call that should be cached?

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 257 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAxU_r7PHRpM09eYwJ:-LDAxU_r7PHRpM09eYwK:b9l0fpd) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L257)):*
> ```PowerShell
>         {
>             Write-Verbose ($LocalizedData.ManagedServiceAccountNotFound -f $ServiceAccountName)
>             Write-Verbose ($LocalizedData.AddingManagedServiceAccount -f $ServiceAccountName)
> ```

Two verbose messages here - it should suffice with one (with a correct localized message).

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 283 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB3ITvAwgzjvZoLtQj:-LDB3ITvAwgzjvZoLtQk:b-sfkvxn) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L283)):*
> ```PowerShell
> 
> ## Import the common AD functions
> $adCommonFunctions = Join-Path -Path (Split-Path -Path $PSScriptRoot -Parent) -ChildPath '\MSFT_xADCommon\MSFT_xADCommon.ps1'
> ```

Please move this row and the next to the top of the file, so it's easier to see that helper functions are loaded.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 4 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB3_8e8KW9QrBk2h6A:-LDB3_8e8KW9QrBk2h6B:bpk2qds) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L4)):*
> ```PowerShell
> param()
> 
> $Global:DSCModuleName      = 'xActiveDirectory' # Example xNetworking
> ```

Remove template comment

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 5 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB3c6_9BaGlxn-eSBv:-LDB3c6_9BaGlxn-eSBw:bpk2qds) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L5)):*
> ```PowerShell
> 
> $Global:DSCModuleName      = 'xActiveDirectory' # Example xNetworking
> $Global:DSCResourceName    = 'MSFT_xADManagedServiceAccount' # Example MSFT_xFirewall
> ```

Remove template comment

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 28 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB3fkP6OiptahPhk_x:-LDB3fkP6OiptahPhk_y:bhzc7rb) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L28)):*
> ```PowerShell
>     #region Pester Tests
> 
>     # The InModuleScope command allows you to perform white-box unit testing on the internal
> ```

Please use comment block for comments over multiple rows.

```
<% 
    Text ...
%>
```

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 56 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB46tK0znXxC0y324k:-LDB46tK0znXxC0y324l:b-3r6r1i) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L56)):*
> ```
> { }
> ```

Not needed if not mocking any code.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 56 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB4H7C6cL3riVJ7tiO:-LDB4H7C6cL3riVJ7tiP:b-za1pnf) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L56)):*
> ```PowerShell
>         #region Function Get-TargetResource
>         Describe "$($Global:DSCResourceName)\Get-TargetResource" {
>             Mock Assert-Module -ParameterFilter { $ModuleName -eq 'ActiveDirectory' } { }
> ```

We should use named parameters `Mock -CommandName Assert-Module -ParameterFiler { ... }`. Throughout. 

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 59 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB4U2939no4ext6B85:-LDB4U2939no4ext6B86:b9d1b6n) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L59)):*
> ```
> Mock Get-ADServiceAccount { return $fakeADMSA }
> ```

We should use named parameters `Mock -CommandName Get-ADServiceAccount -MockWith { return $fakeADMSA }`. Throughout.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 63 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB4r3F8-SGe3yYfZif:-LDB4r3F8-SGe3yYfZig:b-kdiuag) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L63)):*
> ```
> Assert-MockCalled Assert-Module
> ```

We should use named parameters `Assert-MockCalled -CommandName Assert-Module`. Throughout.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 63 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB4xQTC6ejXmT58Ps0:-LDB4xQTC6ejXmT58Ps1:br32e7n) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L63)):*
> ```PowerShell
>                 $null = Get-TargetResource @testPresentParams
> 
>                 Assert-MockCalled Assert-Module -ParameterFilter { $ModuleName -eq 'ActiveDirectory' } -Scope It;
> ```

We could use `-Exactly -Times 1` to make usre we are calling it exactly the right number of times. Throughout.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 66 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB6UbK6aoXZn6vLrvk:-LDB6UbK6aoXZn6vLrvl:b9pphi6) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L66)):*
> ```
> It "Returns
> ```

We should start all it-block with 'Should', for example `"Should return 'Ensure' ...'`.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 66 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB6xPK9SgNdKfH7T_c:-LDB6xPK9SgNdKfH7T_d:b-ttmsm7) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L66)):*
> ```
> group
> ```

The word group seems wrong here. Throughout.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 71 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB5QCJCoqwkBFJpkPc:-LDB5QCK8SvY1aRKqlRy:b62eclh) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L71)):*
> ```PowerShell
>                 (Get-TargetResource @testPresentParams).Ensure | Should Be 'Present';
>             }
> ```

We should also test for it returning all the correct values in the properties in the schema.mof: `It 'Should return the correct values'`, suggest a test when `Context 'When the system is in the desired state'` and `Context 'When the system is not in the desired state'`

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 97 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB8YD_4f458-uL3_nh:-LDB8YD_4f458-uL3_ni:b-sp6x0l) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L97)):*
> ```PowerShell
> 
>         #region Function Test-TargetResource
>         Describe "$($Global:DSCResourceName)\Test-TargetResource" {
> ```

We should test for all states. 

* Ensure -eq 'Present' and in desired state
* Ensure -eq 'Present' and not in desired state
* Ensure -eq 'Absent' and in desired state
* Ensure -eq 'Absent' and not in desired state

Suggest adding the context blocks `Context 'When the system is in the desired state'` and `Context 'When the system is not in the desired state'`.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 100 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB7y4e9xrI1548Bs5x:-LDB7y4e9xrI1548Bs5y:b-a4m2w7) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L100)):*
> ```
> "Fails when MSA does not exist and 'Ensure' is 'Present'"
> ```

The test does not fail (throws) so a better description for this one could be; "Should return $false when MSA does not exist and  'Ensure' is assigned 'Present'"

Throughout the descriptions that start with 'Fails...'

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 162 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDBAdXi4WE717caWoFZ:-LDBAdXi4WE717caWoF_:b-9eqiaj) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L162)):*
> ```PowerShell
>             Mock Assert-Module -ParameterFilter { $ModuleName -eq 'ActiveDirectory' } { }
> 
>             It "Calls 'New-ADServiceAccount' when 'Ensure' is 'Present' and the group does not exist" {
> ```

Cosider using Context-blocks to shorten these description? 

``` 
Context "When 'Ensure' is 'Present' and the group does not exist" {
    It "Should call 'New-ADServiceAccount'" {
        ...
    }
}
```

Throughout.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 177 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB9cUO1jWIAbwMQx84:-LDB9cUO1jWIAbwMQx85:bicvhfp) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L177)):*
> ```PowerShell
>             }
> 
>             foreach ($property in $testProperties.Keys) {
> ```

Consider using -`TestCases` instead of looping in foreach-block. 
https://github.com/pester/Pester/wiki/It#testcases
http://mikefrobbins.com/2016/12/09/loop-through-a-collection-of-items-with-the-pester-testcases-parameter-instead-of-using-a-foreach-loop/

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 188 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDBAC7N9G-rwmobZ3AO:-LDBAC7N9G-rwmobZ3AP:b-e7mun6) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L188)):*
> ```
> -Exactly 1
> ```

`-Exactly -Times 1`

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 192 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDBBEAG2dU-pztkoEew:-LDBBEAG2dU-pztkoEex:b9n0o4q) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L192)):*
> ```
> MSA
> ```

We should avoid abbreviations. Please write out the correct name.

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 220 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDBBgJc2J0LSW9ba6e9:-LDBBgJc2J0LSW9ba6eA:b-3gstka) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L220)):*
> ```
> (#106)
> ```

Suggest we add a comment above this It-block `# Regression test for issue #106`, we could then removed the reference from the It-block description (but okay to keep it too).

---

*[Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1, line 245 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB4ix30upZbT7MhLS6:-LDB4ix30upZbT7MhLS7:bpk2qds) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/Tests/Unit/MSFT_xADManagedServiceAccount.Tests.ps1#L245)):*
> ```PowerShell
>     #endregion
> 
>     # TODO: Other Optional Cleanup Code Goes Here...
> ```

Remove template comment

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/196)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 2 of 4 files at r2.
Review status: all files reviewed at latest revision, 29 unresolved discussions.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 30 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDAzLMe8SGDLY0r69UV:-LDH6hXK3sQJyO5jIRkj:b-jr5v18) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L30)):*
<details><summary><i>Previously, awickham10 (Andrew Wickham) wrote…</i></summary><blockquote>

> We should not have any unused non-mandatory parameters in the Get-TargetResource function. Unused parameters should be removed from the Get-TargetResource.
Is there one in particular that you don't think is being used? I think they all are.
</blockquote></details>

For example we should not return `$DisplayName` in the hash-table when the ;SA is missing, I think we should return `$null` then. Same goes for the others. The properties should be returned to the current value or `$null` if it's missing.  There are exceptions to this, for example if there are a Required parameter that is used to choose code path, that might be returned as the same value as the user assigned in the configuration. 

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 79 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB26e_-hf9UVnoDCi6:-LDH6AKQ9u37sVjuu530:b-evdy49) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L79)):*
<details><summary><i>Previously, awickham10 (Andrew Wickham) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm not seeing this as resolved. :slightly_smiling_face:  Can you explain how you are catching other exceptions than the `ADIdentityNotFoundException`? Or, if you feel it is not necessary, then please explain that. Just so I'm with you in your thinking.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB-1fuAhBcYFep-9lG:-LDH84sk1RwjdEe6Iuic:bu3wpgb) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L82)):*
<details><summary><i>Previously, awickham10 (Andrew Wickham) wrote…</i></summary><blockquote>

Are you thinking something like this? Seems like a ton of extra conditionals to evaluate.

```
    $targetResource = @{}
    if ($adServiceAccount -and $adServiceAccount.Name) 
    {
        $targetResource['ServiceAccountName'] = $adServiceAccount.Name
    }
    else
    {
        $targetResource['ServiceAccountName'] = $Name
    }
```
</blockquote></details>

Sorry for not being clear. I meant something like this, initalize the hash table once.
```
        $targetResource = @{
            ServiceAccountName = $null
            Path = $null
            Description = $null
            DisplayName = $null
            Ensure = 'Absent'
        }
```

Then the logic that finds the MSA do `$targetResource['ServiceAccountName'] = $adServiceAccount.Name` for each property, but if the account is absent it just needs to returns the hash table.
Would that work?

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/196)*
<!-- Sent from Reviewable.io -->

awickham10 commented:
Review status: all files reviewed at latest revision, 29 unresolved discussions.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 79 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB26e_-hf9UVnoDCi6:-LDHFg_-ClTRsDP3bIhL:b8tdr0m) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L79)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

I'm not seeing this as resolved. :slightly_smiling_face:  Can you explain how you are catching other exceptions than the `ADIdentityNotFoundException`? Or, if you feel it is not necessary, then please explain that. Just so I'm with you in your thinking.
</blockquote></details>

I haven't pushed it from my local yet - sorry.

---

*[DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1, line 82 at r1](https://reviewable.io/reviews/powershell/xactivedirectory/196#-LDB-1fuAhBcYFep-9lG:-LDHG-NTEMUI724gj6zA:b2080e2) ([raw file](https://github.com/powershell/xactivedirectory/blob/398e373f8cc9014335979b9ca3aa46f54f17713d/DSCResources/MSFT_xADManagedServiceAccount/MSFT_xADManagedServiceAccount.psm1#L82)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Sorry for not being clear. I meant something like this, initalize the hash table once.
```
        $targetResource = @{
            ServiceAccountName = $null
            Path = $null
            Description = $null
            DisplayName = $null
            Ensure = 'Absent'
        }
```

Then the logic that finds the MSA do `$targetResource['ServiceAccountName'] = $adServiceAccount.Name` for each property, but if the account is absent it just needs to returns the hash table.
Would that work?
</blockquote></details>

I think I got it. That makes sense based on your other comment about unused parameters. I'll see what I can do.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/196)*
<!-- Sent from Reviewable.io -->

kungfu71186 commented:
@awickham10 @johlju Is this still pending review?

johlju commented:
I don’t see a commit after my last comment, so my thought is that I’m still waiting for code fixes. If that is wrong then I will review again.

kungfu71186 commented:
@johlju ok thanks. Maybe I can pick up where @awickham10 left off.  Looks like everything is pretty much there, mostly just clean up

johlju commented:
@kungfu71186 if yu want to continue to work on this, then please do since it is abandoned. See here also https://github.com/PowerShell/DscResources/blob/master/GettingStartedWithGitHub.md#how-to-continue-working-on-a-pull-request-pr-when-an-author-contributor-is-unable-to-complete-it

awickham10 commented:
I have some availability this week to pick back up on this @kungfu71186. I also have some work coming up that will require this functionality.

I think one of the biggest problems for me is that I don't have access to a test AD domain to give it a shot on. Do you think you could help me out with that?

kungfu71186 commented:
@awickham10 yeah of course. I've got plenty of lab space. I was going to write and recommend a managed service resource myself until I found this. So I'm more than willing to help out with this.

awickham10 commented:
I have more work to do on this, just getting everything caught up.

awickham10 commented:
@johlju I think I addressed the issues in the review. Let me know if you think there is still work to be done.

@kungfu71186 if you can give this a try in your lab that would be great.

kungfu71186 commented:
@awickham10 yeah definitely, ill give a shot and let you know

kungfu71186 commented:
@awickham10 @johlju Everything seems to work fine. I tried absent, ensure, path, credential, displayname. It all seems to work very well. 

Few issues though that I see.
1. this one not so big, but just awkward to see. When setting the account to Absent and also setting some other parameter like Path or DisplayName it will check to see if those parameters are correct, but the account doesn't exist so it will never be correct and always show those warnings in verbose mode. Not a big deal, just something a little awkward, but then again, setting absent with other parameters is awkward as well.

2. This is a bigger issue I think. You can now create the managed service accounts, but it's not installed on anything or linked to a computer so wouldn't actually be able to use it. I would still have to go in and manually install it or create another script or resource. I would probably create a new parameter for this. To ensure it's installed on the node. 

Overall this is a good start for managed service accounts. I would proceed with getting this resource added and then adding on the actual install part and also group managed service accounts.

But the service accounts shouldn't be super hard since there is already:

`Install-ADServiceAccount`
Possible Scenarios:
1. Account doesn't exist, so it can't be installed
2. Account is already installed on another computer (Duplicate Backlink) (Which means you need another option to force and overwrite any other links)
3. Account can be installed

`Test-ADServiceAccount`
If account is installed and has proper permissions, should return true. All other scenarios is false.

`Uninstall-ADServiceAccount`

awickham10 commented:
Awesome feedback, thank you!

> 1. this one is not so big, but just awkward to see. When setting the account to Absent and also setting some other parameter like Path or DisplayName it will check to see if those parameters are correct, but the account doesn't exist so it will never be correct and always show those warnings in verbose mode. Not a big deal, just something a little awkward, but then again, setting absent with other parameters is awkward as well.

Do we think the ensure testing should be moved to the start of the Test-TargetResource and return $false immediately if the resource is absent? That would clean up the warnings and save some unnecessary processing.

kungfu71186 commented:
@awickham10  I figured that question would be more for @johlju, but after looking at some other code, there is a check for absent or not. I would probably say, yes and just copy that.
For example: https://github.com/PowerShell/xActiveDirectory/blob/e4817bf4fa710c79455816a56a20665bd0dd1162/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L241

johlju commented:
It seems we have a problem with the tests outside of this PR. Must be a new Pester that making the tests acting up. I will have a look at that. I can merge until the tests passes. :/

johlju commented:
I have merged a PR fixing the tests, so if the PR is rebased the tests that was failing outside of the changes in this PR will pass now.

kungfu71186 commented:
@johlju I have rebased in another fork and fixed the issues in the review. Not sure if this has been abandoned again or not though, but I made a pull request from @awickham10 fork. https://github.com/awickham10/xActiveDirectory/pull/1

Only thing is, I prepped for grouped managed accounts as these will take a lot more testing to implement. Is this acceptable to do? Should I throw an error or something if someone tries to use a Group account right now?

awickham10 commented:
Sorry, not abandoned just wanted to look through what you did and it's been a crazy week. I went ahead and merged your PR in my fork and will look later so I don't hold you up.

kungfu71186 commented:
@awickham10  ok no problem. Fixing the tests cases now. Just making the tests cases for single computer msa work and then ill get some test cases for group.

awickham10 commented:
> @awickham10 ok no problem. Fixing the tests cases now. Just making the tests cases for single computer msa work and then ill get some test cases for group.

Just let me know when you're ready for me to merge.

johlju commented:
@kungfu71186 if there is any issue with gMSA that is not gonna be fixed in this PR (either by you or by Andrew) then please raise an issue so it can be tracked. Then we can leave it out of this PR if you like.

johlju commented:
Let me know when this is ready for review again :)

kungfu71186 commented:
@johlju Yeah definitely. I only know of one problem so far, so I can get the test cases fixed up for gMSA and we can proceed from there. Then it will just be enhancements and bug fixes

kungfu71186 commented:
@johlju Should be ready for review. 

@awickham10 thanks!

johlju commented:
I will review again as soon as I have time. :smiley: Meanwhile, could you rebase this PR against dev? New changes was merged to dev.

kungfu71186 commented:
@johlju It has been rebased and merged. Lot's of changes, so may just want to start over on review.

johlju commented:
@kungfu71186 we should not export the function today, maybe we can call the function `Compare-ResourceState` instead? Or something better that explains what it does, instead of having an X at the end of the function?

johlju commented:
Getting back to reviewing this, last week there was a lot of work at the day job.

kungfu71186 commented:
No worries. I know how that goes. I had a chance to rewrite the test cases at least. Found a few bugs and fixed those

johlju commented:
@kungfu71186 I started to go through the unit tests. Awesome work on those! 

I just added a few comments because I think we can simplify them a bit. We should try to avoid mocking calls in the second level. I think we should only mock on the first level, calls that is made by the function (unit) we are testing. Does that make sense?

*I really miss stubs for these unit tests, I can't run them om my laptop without it, but that is for another PR.* :smile:

johlju commented:
No need to mock Get-ADObjectParentDN, or check if it is called. I thought this was a cmdlet from ActiveDirectory module. :)

kungfu71186 commented:
I didn't know you could do a context in a context. Yeah that would have made it easier 😀

There will be other PRs for this as it's not full featured yet, so I can fix the readability of the test cases at that time.

johlju commented:
Pester v4.x made it possible to have nested context-blocks. :)

I will merge this when I get back to the computer.
